pkg/tbtc/inactivity.go

```diff
@@ -0,0 +1,451 @@
+package tbtc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+
+	"github.com/ipfs/go-log/v2"
+	"go.uber.org/zap"
+	"golang.org/x/sync/semaphore"
+
+	"github.com/keep-network/keep-core/pkg/bitcoin"
+	"github.com/keep-network/keep-core/pkg/chain"
+	"github.com/keep-network/keep-core/pkg/generator"
+	"github.com/keep-network/keep-core/pkg/net"
+	"github.com/keep-network/keep-core/pkg/protocol/group"
+	"github.com/keep-network/keep-core/pkg/protocol/inactivity"
+)
+
+const (
+	// inactivityClaimSubmissionDelayStepBlocks determines the delay step in blocks
+	// that is used to calculate the submission delay period that should be respected
+	// by the given member to avoid all members submitting the same inactivity claim
+	// at the same time.
+	inactivityClaimSubmissionDelayStepBlocks = 2
+)
+
+// errInactivityClaimExecutorBusy is an error returned when the inactivity claim
+// executor cannot execute the inactivity claim due to another inactivity claim
+// execution in progress.
+var errInactivityClaimExecutorBusy = fmt.Errorf("inactivity claim executor is busy")
+
+type inactivityClaimExecutor struct {
+	lock *semaphore.Weighted
+
+	chain               Chain
+	signers             []*signer
+	broadcastChannel    net.BroadcastChannel
+	membershipValidator *group.MembershipValidator
+	groupParameters     *GroupParameters
+	protocolLatch       *generator.ProtocolLatch
+
+	waitForBlockFn waitForBlockFn
+}
+
+func newInactivityClaimExecutor(
+	chain Chain,
+	signers []*signer,
+	broadcastChannel net.BroadcastChannel,
+	membershipValidator *group.MembershipValidator,
+	groupParameters *GroupParameters,
+	protocolLatch *generator.ProtocolLatch,
+	waitForBlockFn waitForBlockFn,
+) *inactivityClaimExecutor {
+	return &inactivityClaimExecutor{
+		lock:                semaphore.NewWeighted(1),
+		chain:               chain,
+		signers:             signers,
+		broadcastChannel:    broadcastChannel,
+		membershipValidator: membershipValidator,
+		groupParameters:     groupParameters,
+		protocolLatch:       protocolLatch,
+		waitForBlockFn:      waitForBlockFn,
+	}
+}
+
+func (ice *inactivityClaimExecutor) claimInactivity(
+	parentCtx context.Context,
```

Comment:
Let's simplify it and rename the child context (see next comment):
```suggestion
	ctx context.Context,
```